File: packages/scelo-py/src/scelo/_table.py
"""Table: a pandas DataFrame that carries its own caveats.

Every Scelo output that is table-shaped is a :class:`Table`: a real
``pandas.DataFrame`` (slice it, merge it, plot it, ``.to_csv`` it) with four
extra attributes that survive most pandas operations:

``title``       one line naming the table,
``notes``       the things an actuary should know before trusting it,
``basis``       a one-line provenance label ("Gompertz–Makeham (illustrative) · i = 4 %"),
``provenance``  a dict stamped by :func:`scelo.hard`: content hash, time,
                scelo version, the audit entries that produced it.

Printing a Table prints the frame and then its notes, so the caveat is on
screen at the moment the number is, the one-way pipeline rule in practice:
a number never travels without its basis.
"""

from __future__ import annotations

from typing import Any, Dict, List, Optional

import pandas as pd

_NOTE_PREFIX = "  · "


class Table(pd.DataFrame):
    """A DataFrame with ``title``, ``notes``, ``basis`` and ``provenance``."""

    _metadata = ["title", "notes", "basis", "provenance", "stage"]

    def __init__(
        self,
        data=None,
        *args: Any,
        title: Optional[str] = None,
        notes: Optional[List[str]] = None,
        basis: Optional[str] = None,
        provenance: Optional[Dict[str, Any]] = None,
        stage: Optional[str] = None,
        **kwargs: Any,
    ) -> None:
        super().__init__(data, *args, **kwargs)
        # Preserve metadata when constructed from another Table.
        src = data if isinstance(data, Table) else None
        has = src is not None
        object.__setattr__(self, "title", title if title is not None else (src.title if has else None))
        object.__setattr__(self, "notes", list(notes) if notes is not None else (list(src.notes) if has else []))
        object.__setattr__(self, "basis", basis if basis is not None else (src.basis if has else None))
        object.__setattr__(
            self, "provenance", dict(provenance) if provenance is not None else (dict(src.provenance) if has else {})
        )
        object.__setattr__(self, "stage", stage if stage is not None else (src.stage if has else "soft"))

    # pandas subclassing contract ------------------------------------------------
    @property
    def _constructor(self):
        return Table

    @property
    def _constructor_sliced(self):
        return pd.Series

    # presentation ----------------------------------------------------------------
    def _footer(self) -> str:
        lines: List[str] = []
        if self.title:
            lines.append(f"— {self.title}")
        if self.basis:
            lines.append(f"  basis: {self.basis}")
        for n in self.notes or []:
            lines.append(_NOTE_PREFIX + n)
        if self.provenance:
            h = self.provenance.get("sha256", "")
            if h:
                lines.append(f"  hard · {h[:12]} · scelo {self.provenance.get('scelo', '?')} · {self.provenance.get('at', '')}")
        return "\n".join(lines)

    def __repr__(self) -> str:  # pragma: no cover - presentation
        body = super().__repr__()
        foot = self._footer()
        return body + ("\n" + foot if foot else "")

    def _repr_html_(self) -> Optional[str]:  # pragma: no cover - presentation
        body = super()._repr_html_()
        foot = self._footer()
        if not foot or body is None:
            return body
        esc = foot.replace("&", "&amp;").replace("<", "&lt;").replace(">", "&gt;")
        return body + f"<pre style='font-size:0.85em;opacity:0.8'>{esc}</pre>"

    # convenience -------------------------------------------------------------------
    def note(self, text: str) -> "Table":
        """Append a note (returns self, for chaining)."""
        self.notes.append(text)
        return self

    @property
    def df(self) -> pd.DataFrame:
        """A plain ``pandas.DataFrame`` copy (drops the Scelo metadata)."""
        return pd.DataFrame(self)

    def to_markdown_report(self) -> str:
        """Title, basis, the table (markdown) and the notes, one block."""
        out: List[str] = []
        if self.title:
            out.append(f"### {self.title}")
        if self.basis:
            out.append(f"*Basis:* {self.basis}")
        out.append("")
        try:
            out.append(pd.DataFrame(self).to_markdown(index=False))
        except ImportError:  # tabulate not installed
            out.append("```\n" + pd.DataFrame(self).to_string(index=False) + "\n```")
        if self.notes:
            out.append("")
            out.extend(f"- {n}" for n in self.notes)
        if self.provenance:
            out.append("")
            out.append(
                f"<sub>hard · sha256 {self.provenance.get('sha256', '')[:16]} · "
                f"scelo {self.provenance.get('scelo', '')} · {self.provenance.get('at', '')}</sub>"
            )
        return "\n".join(out)


def as_table(
    df: pd.DataFrame,
    title: Optional[str] = None,
    notes: Optional[List[str]] = None,
    basis: Optional[str] = None,
    stage: str = "hard",
) -> Table:
    """Wrap a DataFrame as a :class:`Table` (copying metadata if it already is one)."""
    t = Table(df)
    if title is not None:
        object.__setattr__(t, "title", title)
    if notes is not None:
        object.__setattr__(t, "notes", list(notes))
    if basis is not None:
        object.__setattr__(t, "basis", basis)
    object.__setattr__(t, "stage", stage)
    return t


def notes(x: Any) -> List[str]:
    """The notes attached to a Scelo result (empty list for plain objects)."""
    return list(getattr(x, "notes", []) or [])

File: packages/scelo-r/R/table.R
# A scelo_table is a data.frame that carries its own caveats.
#
# Every table-shaped Scelo result is one: a plain data.frame (subset it,
# merge it, plot it) with four extra attributes that survive the common
# operations: title, basis (one line of provenance), notes (what an actuary
# should know before trusting it) and provenance (the stamp sc_hard() adds:
# content hash, time, scelo version, audit trail). Printing shows the frame
# and then the notes, so the caveat is on screen at the moment the number is.

#' Make a scelo table
#'
#' Wrap a data frame as a `scelo_table`, the type every Scelo result uses: a
#' data.frame carrying `title`, `basis`, `notes`, `provenance` and `stage`.
#'
#' @param df A data frame.
#' @param title One line naming the table.
#' @param basis One line of provenance ("Gompertz-Makeham (illustrative) · i = 4 %").
#' @param notes Character vector of caveats.
#' @param stage "soft" or "hard".
#' @param provenance A list written by [sc_hard()].
#' @return A `scelo_table`.
#' @examples
#' t <- sc_table(data.frame(x = 1:3), title = "three rows", notes = "toy")
#' print(t)
#' @export
sc_table <- function(df, title = NULL, basis = NULL, notes = character(), stage = "soft", provenance = NULL) {
  df <- as.data.frame(df, stringsAsFactors = FALSE)
  if (inherits(df, "scelo_table")) {
    if (is.null(title)) title <- attr(df, "title")
    if (is.null(basis)) basis <- attr(df, "basis")
    if (!length(notes)) notes <- attr(df, "notes")
    if (is.null(provenance)) provenance <- attr(df, "provenance")
  }
  attr(df, "title") <- title
  attr(df, "basis") <- basis
  attr(df, "notes") <- as.character(notes %||% character())
  attr(df, "provenance") <- provenance
  attr(df, "stage") <- stage
  class(df) <- unique(c("scelo_table", class(df)))
  df
}

`%||%` <- function(a, b) if (is.null(a)) b else a

#' Notes, title and basis of a Scelo result
#'
#' @param x Any object; a `scelo_table` carries notes.
#' @return `sc_notes()` a character vector (empty for plain objects);
#'   `sc_title()` / `sc_basis()` a string or `NULL`.
#' @examples
#' sc_notes(sc_life_table())
#' @export
sc_notes <- function(x) as.character(attr(x, "notes") %||% character())

#' @rdname sc_notes
#' @export
sc_title <- function(x) attr(x, "title")

#' @rdname sc_notes
#' @export
sc_basis <- function(x) attr(x, "basis")

#' Append a note to a table
#'
#' @param x A `scelo_table`.
#' @param text The note.
#' @return The table, with the note appended.
#' @export
sc_note <- function(x, text) {
  attr(x, "notes") <- c(sc_notes(x), text)
  x
}

#' Plain data frame
#'
#' Drop the Scelo attributes and return an ordinary data.frame.
#' @param x A `scelo_table`.
#' @return A data.frame.
#' @export
sc_df <- function(x) {
  for (a in c("title", "basis", "notes", "provenance", "stage")) attr(x, a) <- NULL
  class(x) <- "data.frame"
  x
}

.sc_footer <- function(x) {
  out <- character()
  if (!is.null(attr(x, "title"))) out <- c(out, paste0("— ", attr(x, "title")))
  if (!is.null(attr(x, "basis"))) out <- c(out, paste0("  basis: ", attr(x, "basis")))
  for (n in sc_notes(x)) out <- c(out, paste0("  · ", n))
  p <- attr(x, "provenance")
  if (!is.null(p) && !is.null(p$hash)) out <- c(out, sprintf("  hard · %s · scelo %s · %s", substr(p$hash, 1, 12), p$scelo, p$at))
  out
}

#' @export
print.scelo_table <- function(x, ..., n = getOption("scelo.print_rows", 60L)) {
  df <- sc_df(x)
  if (nrow(df) > n) {
    print(utils::head(df, n), ...)
    cat(sprintf("  ... %d more rows (options(scelo.print_rows = N) to show more)\n", nrow(df) - n))
  } else {
    print(df, ...)
  }
  foot <- .sc_footer(x)
  if (length(foot)) cat(foot, sep = "\n")
  invisible(x)
}

#' @export
`[.scelo_table` <- function(x, i, j, ..., drop = FALSE) {
  keep <- c("title", "basis", "notes", "provenance", "stage")
  attrs <- lapply(keep, function(a) attr(x, a))
  out <- NextMethod()
  if (is.data.frame(out)) {
    for (k in seq_along(keep)) attr(out, keep[k]) <- attrs[[k]]
    class(out) <- unique(c("scelo_table", class(out)))
  }
  out
}

#' @export
as.data.frame.scelo_table <- function(x, ...) sc_df(x)

#' Format a table as a Markdown report block
#'
#' Title, basis, a pipe table and the notes, the block [sc_report()] writes
#' for each table.
#' @param x A data frame or `scelo_table`.
#' @param digits Significant digits for numbers.
#' @return A single string.
#' @export
sc_markdown <- function(x, digits = 6) {
  df <- if (inherits(x, "scelo_table")) sc_df(x) else as.data.frame(x)
  out <- character()
  if (!is.null(attr(x, "title"))) out <- c(out, paste0("### ", attr(x, "title")))
  if (!is.null(attr(x, "basis"))) out <- c(out, paste0("*Basis:* ", attr(x, "basis")))
  out <- c(out, "", .sc_pipe_table(df, digits))
  if (length(sc_notes(x))) out <- c(out, "", paste0("- ", sc_notes(x)))
  p <- attr(x, "provenance")
  if (!is.null(p) && !is.null(p$hash)) out <- c(out, "", sprintf("<sub>hard · %s · scelo %s · %s</sub>", substr(p$hash, 1, 16), p$scelo, p$at))
  paste(out, collapse = "\n")
}

.sc_fmt_cell <- function(v, digits) {
  if (is.numeric(v)) ifelse(is.na(v), "", formatC(v, digits = digits, format = "g", big.mark = ",")) else ifelse(is.na(v), "", as.character(v))
}

.sc_pipe_table <- function(df, digits = 6) {
  if (!nrow(df)) return("(empty)")
  cells <- vapply(names(df), function(c) .sc_fmt_cell(df[[c]], digits), character(nrow(df)))
  if (is.null(dim(cells))) cells <- matrix(cells, nrow = 1)
  header <- paste0("| ", paste(names(df), collapse = " | "), " |")
  sep <- paste0("|", paste(rep("---", ncol(df)), collapse = "|"), "|")
  body <- apply(cells, 1, function(r) paste0("| ", paste(r, collapse = " | "), " |"))
  c(header, sep, body)
}
